### PR TITLE
feat: restoring applied filters with search params

### LINKED ISSUE #24: Restore filters on page load
Currently, the filters are lost once we reload the activitylog page or if we accidentally close the activitylog page. But we want the filters to be applied (that we set previously) on activitylog page loads.

### PR DESCRIPTION
**Pull request by atiqueahmedziad**
Fixes #24 

This PR updates URL with filter information as soon as any filter is applied/ removes. This enables the user to restore the filters via URL.

### CODE REVIEW

File: src/lib/ext-activitylog.js [hunk 1]
```diff
@@ -264,6 +283,37 @@ class Controller {
         }
       });
 
+      const ids = filterIds?.split(',');
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: This is done, I have used `JSON.stringify` and `JSON.parse` :-)

File: src/lib/ext-activitylog.js [hunk 2]
```diff
@@ -264,6 +283,37 @@ class Controller {
...
+
+      const viewTypes = filterViewTypes?.split(',');
+      // converting empty string to undefined
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: Thanks :-) Done.

File: src/lib/ext-activitylog.js [hunk 3]
```diff
@@ -221,12 +232,20 @@ class Controller {
...
+    const filterIds = searchParams.get('ids');
+    const filterViewTypes = searchParams.get('viewTypes');
+    const filterTypes = searchParams.get('types');
+    const filterNames = searchParams.get('names');
     const filterTabId = parseInt(searchParams.get('filterTabId'), 10);
+    const filterKeyword = searchParams.get('keyword');
+    const filterTimestamp = searchParams.get('timestamp');
 
-    if (fileName) {
+    if (loadedFileName) {
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js:240]: yes :-) Fixed it.

File: src/lib/ext-activitylog.js [hunk 4]
```diff
@@ -336,14 +386,32 @@ class Controller {
...
+    const typesQuery = `types=${[...type].toString()}`;
+    const namesQuery = `names=${[...name].toString()}`;
+    const timestampsQuery = `timestamp=${startTime},${stopTime}`;
+    const keywordQuery = `keyword=${keyword}`;
+    const tabIdQuery = `filterTabId=${tabId}`;
+
+    history.replaceState(
+      null,
+      null,
+      `?${idsQuery}&${viewTypesQuery}&${typesQuery}&${namesQuery}&${keywordQuery}&${tabIdQuery}&${timestampsQuery}`
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: Used `URLSearchParams` and moved the serialization and de-serialization logic in `formatters.js`

File: src/lib/web-component/filter-option/filter-option-element.js [hunk 5]
```diff
@@ -35,11 +37,21 @@ export class FilterOption extends HTMLElement {
...
+
+    this.dispatchFilterChangeEvent();
+
+    // When any new unknown filter option is found, it is enabled by default.
+    this.isCheckboxChecked = () => true;
+  }
+
+  setFilterFromURL(searchParamLabels) {
+    if (searchParamLabels.size > 0) {
+      this.isCheckboxChecked = (checkboxLabel) =>
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/web-component/filter-option/filter-option-element.js]: done :-) Currently I am storing the unchecked filter options in `filter` object.

File: src/lib/ext-activitylog.js [hunk 6]
```diff
@@ -264,6 +283,37 @@ class Controller {
...
+        const updateFilter = {
+          id: new Set(ids),
+          viewType: new Set(viewTypes),
+          type: new Set(types),
+          name: new Set(names),
+          keyword: filterKeyword,
+          timeStamp: timestamps,
+        };
+
+        this.view.setFilterOptionsFromURL(updateFilter);
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js:283]: Now, the flow is URL params -> Controller -> Model -> View ...

**Review by Rob--W — COMMENTED** [review]
Review not complete yet, but submitting so we can discuss it during the current meeting.

File: src/lib/web-component/filter-option/filter-option-element.js [hunk 7]
```diff
@@ -6,8 +6,8 @@ export class FilterOption extends HTMLElement {
     super();
 
     this.viewCheckboxLabels = new Set();
-    // It contains the checked checkboxes
-    this.activeCheckboxLabels = new Set();
+    // It contains the unchecked checkboxes
+    this.inActiveCheckboxLabels = new Set();
```
Comment (Rob--W) [inline src/lib/web-component/filter-option/filter-option-element.js:10]: Rename to `inactive`, `inActiveCheckboxLabels` looks like "in active checkbox labels", which doesn't reflect the actual purpose.

File: src/lib/ext-activitylog.js [hunk 8]
```diff
@@ -183,6 +184,17 @@ class View {
     this.apiNameFilter.updateFilterCheckboxes(filteredLogs);
   }
 
+  setFilterOptionsFromURL(updateFilter) {
+    const { id, viewType, type, name, keyword, timeStamp } = updateFilter;
+
+    this.extFilter.setFilterFromURL(id);
```
Comment (Rob--W) [inline src/lib/ext-activitylog.js:190]: `setFilterOptionsFromURL` and `setFilterFromURL` are independent of the URL. For example, if we ever store the filter in a file in the future, the same logic can be used. Please choose a meaningful than that is generic enough to serve the current use case and this potential future use case.

**Review by Rob--W — COMMENTED** [review]
After following the previous suggestion of making the `serialize` / `deserialize` methods stateless / independent of external state, it should be much easier to write unit tests to verify that those two methods work as intended. Please include these tests in the PR, and try to think of edge cases (e.g. incorrect/missing input).

File: src/devtools/devtools.js [hunk 9]
```diff
@@ -3,5 +3,5 @@ const tabId = browser.devtools.inspectedWindow.tabId;
 browser.devtools.panels.create(
   'Extension Activity',
   '/icons/eam-48.png',
-  `/activitylog/activitylog.html?filterTabId=${tabId}`
+  `/activitylog/activitylog.html?tabId=${tabId}`
```
Comment (atiqueahmedziad) [inline src/devtools/devtools.js:6]: To make it consistent with the `filter` object property i.e. `tabId`

File: src/lib/ext-activitylog.js (hunk 8 above)
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: I renamed it to `setExistedFilters` since these filters were set before.

File: src/lib/web-component/filter-option/filter-option-element.js (hunk 7 above)
Comment (atiqueahmedziad) [reply src/lib/web-component/filter-option/filter-option-element.js]: done :-)

File: src/lib/ext-activitylog.js [hunk 10]
```diff
@@ -221,12 +233,13 @@ class Controller {
     const searchParams = new URLSearchParams(
       document.location.search.substring(1)
     );
-    const fileName = searchParams.get('file');
+
+    const loadedFileName = searchParams.get('file');
     const filterTabId = parseInt(searchParams.get('filterTabId'), 10);
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: done :-)

File: src/lib/formatters.js [hunk 11]
```diff
@@ -9,3 +9,69 @@ export function dateTimeFormat(timestamp, options) {
...
+
+    if (
+      key === 'id' ||
+      key === 'viewType' ||
+      key === 'type' ||
+      key === 'name'
+    ) {
+      searchParamVal = JSON.stringify([...updateFilter[key].exclude]);
+    } else if (key === 'timeStamp') {
+      searchParamVal = timestampFormat(updateFilter[key]);
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/formatters.js]: Sorry, Fixed it :-)

File: src/lib/formatters.js [hunk 12]
```diff
@@ -9,3 +9,69 @@ export function dateTimeFormat(timestamp, options) {
...
+
+function timestampFormat(timeStamp) {
+  const startTime = timeStamp?.start ? timeStamp.start : null;
+  const stopTime = timeStamp?.stop ? timeStamp.stop : null;
+  return JSON.stringify([JSON.stringify(startTime), JSON.stringify(stopTime)]);
+}
+
+export function serializeFilters(updateFilter) {
+  const currentURL = new URL(document.location.href);
+  const searchParams = new URLSearchParams(currentURL.search.slice(1));
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/formatters.js]: Yes done now :-)

File: src/lib/formatters.js [hunk 13]
```diff
@@ -9,3 +9,69 @@ export function dateTimeFormat(timestamp, options) {
...
+    } else if (key === 'timeStamp') {
+      searchParamVal = timestampFormat(updateFilter[key]);
+    } else {
+      searchParamVal = updateFilter[key];
+    }
+
+    searchParams.set(key, searchParamVal);
+  });
+
+  history.replaceState(null, null, `?${searchParams}`);
```
[the remark this answers is not in the document]
Comment (atiqueahmedziad) [reply src/lib/formatters.js]: done :-)

File: src/lib/ext-activitylog.js [hunk 14]
```diff
@@ -264,17 +276,25 @@ class Controller {
...
+
       const existingLogs = await this.getExistingLogs();
 
       if (existingLogs.length) {
         this.handleNewLogs(existingLogs);
       }
 
-      if (filterTabId) {
-        this.view.renderHeading({ filterTabId });
+      if (this.model.filter.tabId != null) {
```
Comment (Rob--W) [inline src/lib/ext-activitylog.js:291]: This logic would better be placed in `setExistedFilters`.

File: src/lib/formatters.js [hunk 15]
```diff
@@ -9,3 +9,67 @@ export function dateTimeFormat(timestamp, options) {
...
+  const filterViewTypes = searchParams.get('viewType');
+  const filterTypes = searchParams.get('type');
+  const filterNames = searchParams.get('name');
+  const filterKeyword = searchParams.get('keyword');
+  const filterTimestamp = searchParams.get('timeStamp');
+  const filterTabId = parseInt(searchParams.get('tabId'), 10) || null;
+
+  const ids = getJSONParseVal(filterIds);
+  let viewTypes = getJSONParseVal(filterViewTypes);
+  // converting empty string to undefined
```
Comment (Rob--W) [inline src/lib/formatters.js:55]: This comment isn't accurate any more. Remove it.

File: src/lib/web-component/filter-option/filter-option-element.js [hunk 16]
```diff
@@ -6,8 +6,8 @@ export class FilterOption extends HTMLElement {
     super();
 
     this.viewCheckboxLabels = new Set();
-    // It contains the checked checkboxes
-    this.activeCheckboxLabels = new Set();
+    // It contains the unchecked checkboxes
+    this.inactive = new Set();
```
Comment (Rob--W) [inline src/lib/web-component/filter-option/filter-option-element.js:10]: This name is too vague. If you choose a meaningful name, the comment wouldn't be needed.
How about `uncheckedCheckboxLabels`?

File: src/lib/formatters.js [hunk 17]
```diff
@@ -9,3 +9,67 @@ export function dateTimeFormat(timestamp, options) {
...
+    searchParams.set(key, searchParamVal);
+  });
+
+  return searchParams;
+}
+
+function getJSONParseVal(str) {
+  try {
+    return JSON.parse(str);
+  } catch (error) {
```
Comment (Rob--W) [inline src/lib/formatters.js:39]: I suppose that you've created this to work around errors. Did you mean to include (new?) unit tests that trigger the error, or is that part of follow-up work?

Comment (atiqueahmedziad) [reply src/lib/formatters.js:39]: Since you said to think about the edge cases (e.g. incorrect/missing input), I just put random search params in url and got an error from `JSON.parse`. This is the work arround for that. 
Should I add tests for this?

File: src/lib/web-component/filter-option/filter-option-element.js (hunk 16 above)
Comment (atiqueahmedziad) [reply src/lib/web-component/filter-option/filter-option-element.js:10]: Yeah, `uncheckedCheckboxLabels` sounds great. I also removed the comment.

File: src/lib/formatters.js (hunk 15 above)
Comment (atiqueahmedziad) [reply src/lib/formatters.js:55]: done :-)

File: src/lib/ext-activitylog.js (hunk 14 above)
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js:291]: done :-)

File: src/lib/formatters.js (hunk 17 above)
Comment (Rob--W) [reply src/lib/formatters.js:39]: Yes. There are more ways to break the functionality (e.g. something that is valid JSON but not having the expected type), but you'll discover that in tests.

File: src/lib/ext-activitylog.js [hunk 18]
```diff
@@ -183,6 +184,34 @@ class View {
     this.apiNameFilter.updateFilterCheckboxes(filteredLogs);
   }
 
+  setExistedFilters(updateFilter) {
```
Comment (Rob--W) [inline src/lib/ext-activitylog.js:187]: I meant to comment about this before; You probably meant "set existing filters" instead of "set existed filters".

`setInitialFilter` may be a better name though, because it does seem that the method can be called only once (as the implementations are not yet prepared to deal with setting the filter after the initialization).

File: src/lib/formatters.js [hunk 19]
```diff
@@ -9,3 +9,66 @@ export function dateTimeFormat(timestamp, options) {
...
+    return null;
+  }
+}
+
+export function deSerializeFilters(searchParams) {
+  const filterIds = searchParams.get('id');
+  const filterViewTypes = searchParams.get('viewType');
+  const filterTypes = searchParams.get('type');
+  const filterNames = searchParams.get('name');
+  const filterKeyword = searchParams.get('keyword');
```
Comment (Rob--W) [inline src/lib/formatters.js:49]: `filterKeyword = ... || '';`

Do add a test. To find relevant edge cases, look at your code and try to see if the end result is guaranteed to have the expected type as defined in the Model at `filter`.

File: src/lib/web-component/filter-timestamp/filter-timestamp-element.js [hunk 20]
```diff
@@ -53,6 +53,30 @@ export class FilterTimestamp extends HTMLElement {
...
+    }
+
+    if (timestamp?.stop != null) {
+      this.timeStamp.stop = timestamp.stop;
+      this.stopTimeLabel.textContent = dateTimeFormat(timestamp.stop);
+      this.filterContainer.hidden = false;
+      this.clearStopTimeBtn.hidden = false;
+    }
+
+    this.dispatchFilterChange();
```
Comment (Rob--W) [inline src/lib/web-component/filter-timestamp/filter-timestamp-element.js:77]: Why is it necessary to call `dispatchFilterChange` here?

Comment (atiqueahmedziad) [reply src/lib/web-component/filter-timestamp/filter-timestamp-element.js]: not needed. Updated :-)

File: src/lib/formatters.js (hunk 19 above)
Comment (atiqueahmedziad) [reply src/lib/formatters.js]: Updated and added test that guarantees the same type as defined in the model.

File: src/lib/ext-activitylog.js (hunk 18 above)
Comment (atiqueahmedziad) [reply src/lib/ext-activitylog.js]: Thanks. Updated the name as `setInitialFilter` :-)

File: tests/formatters.test.js [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    timeStamp,
+  };
+
+  const serializedFormat = formatters.serializeFilters(
+    searchParams,
+    updateFilter
+  );
+
+  expect(serializedFormat.get('id')).toEqual(JSON.stringify(extIds));
+  expect(serializedFormat.get('viewType')).toEqual(JSON.stringify(viewTypes));
```
Comment (Rob--W) [inline tests/formatters.test.js:31]: Hardcode the string instead of using `JSON.stringify`. It's not obvious from the code to those unfamiliar with the exact working of `JSON.stringify` that the stringified value is `["popup",null]`.

File: tests/formatters.test.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+  const searchParams = new URLSearchParams(testUrl.search.slice(1));
+  const filterObject = formatters.deSerializeFilters(searchParams);
+
+  expect(filterObject).toMatchObject(expectedFilterObject);
+});
+
+test('deSerializeFilters method should return the default filter object if there are no valid filters', () => {
+  const testUrl = new URL(
+    'http://test.html?invalidParam="wrongValue"&id=null&viewType="wrong-type"&timeStamp="wrong-type"&tabId=undefined'
```
Comment (Rob--W) [inline tests/formatters.test.js:62]: The question is - does one invalid filter disable all other filters? You need to test a filter that includes at least one valid filter to check the difference.

And the test is not exhaustive, you need a test case with the filter value `{}` because it would show an issue in the implementation.

File: tests/formatters.test.js (hunk 21 above)
Comment (atiqueahmedziad) [reply tests/formatters.test.js]: Yeah, done :-)

File: tests/formatters.test.js (hunk 22 above)
Comment (atiqueahmedziad) [reply tests/formatters.test.js]: I have updated this test, now the test description is - `deSerializeFilters method should return the filter object with valid filter properties, ignoring the invalid filter properties`. Is this okay?

File: tests/formatters.test.js [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+  const serializedFormat = formatters.serializeFilters(
+    searchParams,
+    updateFilter
+  );
+
+  const expectedExtIds = '["addon1@test.com","addon2@test.com"]';
+  const expectedViewTypes = '["popup",null]';
+  const expectedTypes = '["api_call","content_script"]';
+  const expectedNames = '["test@api"]';
+  const expectedTabId = '22';
```
Comment (Rob--W) [inline tests/formatters.test.js:34]: This should have been `22`.

Use toStrictEqual instead of isEqual in your assertions to catch this kind of error.

File: tests/formatters.test.js [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    name: { exclude: new Set(['tabs.executeScript']) },
+    tabId: 982,
+    keyword: 'undefined',
+    timeStamp: { start: 123123, stop: 456456 },
+  };
+
+  const searchParams = new URLSearchParams(testUrl.search.slice(1));
+  const filterObject = formatters.deSerializeFilters(searchParams);
+
+  expect(filterObject).toMatchObject(expectedFilterObject);
```
Comment (Rob--W) [inline tests/formatters.test.js:63]: I'm not sure, can you verify that .toMatchObject actually checks the contents of the Set? That is, if you intentionally make the expectation invalid, does the test fail as expected?

File: src/lib/formatters.js [hunk 25]
```diff
@@ -9,3 +9,73 @@ export function dateTimeFormat(timestamp, options) {
...
+    return null;
+  }
+
+  try {
+    const parsedData = JSON.parse(str);
+    if (options?.returnArray) {
+      return Array.isArray(parsedData) ? parsedData : [];
+    } else if (options?.returnObject) {
+      return typeof parsedData === 'object' ? parsedData : null;
+    }
```
Comment (Rob--W) [inline src/lib/formatters.js:47]: return value in other cases is missing.

**Review by Rob--W — APPROVED** [review]
Address the issues below and it's good to merge.

File: tests/formatters.test.js (hunk 23 above)
Comment (atiqueahmedziad) [reply tests/formatters.test.js:34]: I am getting the `"22"` since I am getting the tab id from search params. Should I parse that into number and then check ?

File: tests/formatters.test.js (hunk 24 above)
Comment (atiqueahmedziad) [reply tests/formatters.test.js:63]: Yes, it checks the content of the Set too. :-)

File: src/lib/formatters.js (hunk 25 above)
Comment (atiqueahmedziad) [reply src/lib/formatters.js:49]: added now :-)

File: tests/formatters.test.js (hunk 23 above)
Comment (Rob--W) [reply tests/formatters.test.js]: I don't think that you're getting `"22"` but 22. You're parsing `tabId` before returning it:

```javascript
const filterTabId = parseInt(searchParams.get('tabId'), 10) || null;
```

**Review by Rob--W — APPROVED** [review]

File: tests/formatters.test.js (hunk 23 above)
Comment (atiqueahmedziad) [reply tests/formatters.test.js]: > I don't think that you're getting `"22"` but 22. You're parsing `tabId` before returning it:
> 
> ```js
> const filterTabId = parseInt(searchParams.get('tabId'), 10) || null;
> ```

Yes, I have added this parsing now.
